tsdb/engine/tsm1/engine.go

```diff
@@ -2544,21 +2541,22 @@ func SeriesFieldKeyBytes(seriesKey, field string) []byte {
 	return b
 }
 
-func tsmFieldTypeToInfluxQLDataType(typ byte) (influxql.DataType, error) {
-	switch typ {
-	case BlockFloat64:
-		return influxql.Float, nil
-	case BlockInteger:
-		return influxql.Integer, nil
-	case BlockUnsigned:
-		return influxql.Unsigned, nil
-	case BlockBoolean:
-		return influxql.Boolean, nil
-	case BlockString:
-		return influxql.String, nil
-	default:
-		return influxql.Unknown, fmt.Errorf("unknown block type: %v", typ)
+var (
+	blockToFieldType = []influxql.DataType{
+		BlockFloat64:  influxql.Float,
+		BlockInteger:  influxql.Integer,
+		BlockBoolean:  influxql.Boolean,
+		BlockString:   influxql.String,
+		BlockUnsigned: influxql.Unsigned,
+	}
+)
+
+func BlockTypeToInfluxQLDataType(typ byte) influxql.DataType {
+	if int(typ) < len(blockToFieldType) {
```

Comment:
As we're on an über optimisation drive here, you could also have:

```go
fieldTypesN byte = len(blockToFieldType)
...
...
...
func BlockTypeToInfluxQLDataType(typ byte) influxql.DataType {
    if typ < fieldTypesN {
```

Reply:
I tried that, but alas the bounds check elimination (BCE) optimization can't be used. We can only declare `FieldTypesN` as a `var` and the compiler cannot prove this value doesn't change outside the `BlockTypeToInfluxQLDataType` function.

Reply:
Interesting. 👍